Created Dockerfile for golang 1.7 and 1.8.

yongni opened the pull request:
…ort for runtime specification in build_interop_image.sh

grpc-testing commented:
```
No significant performance differences
```

jtattermusch commented on tools/dockerfile/interoptest/grpc_interop_go/Dockerfile_1.7:
```diff
@@ -0,0 +1,51 @@
+# Copyright 2015, Google Inc.
```
filename should be just `Dockerfile`, that's a standard for docker.
grpc_interop_go1.7  or grpc_interop_go1_7 or sth in that sense

yongni replied:
Do you mean a grpc_interop_go/grpc_interop_go1.7 as the filename or grpc_interop_go1.7/Dockerfile as separate directory?

jtattermusch commented on tools/dockerfile/interoptest/grpc_interop_go/Dockerfile_1.8:
```diff
@@ -0,0 +1,51 @@
+# Copyright 2015, Google Inc.
```
Filename: see above

jtattermusch commented on tools/dockerfile/interoptest/grpc_interop_java/Dockerfile_open7:
```diff
@@ -0,0 +1,69 @@
+# Copyright 2017, Google Inc.
```
Filename: see above

jtattermusch replied:
also `openjdk7` is much more readable than `open7`.

jtattermusch commented on tools/dockerfile/interoptest/grpc_interop_java/Dockerfile_oracle8:
```diff
@@ -0,0 +1,73 @@
+# Copyright 2015, Google Inc.
```
For all dockerfiles:

We actually use templates to generate our docker files to make sure slightly different versions of our dockerfiles stay in sync (e.g.   openjdk8 and oracle8 should be almost identical, with just the installed java package being different - and a template should be used to make sure that happens).

See templates here:
https://github.com/grpc/grpc/tree/master/templates/tools/dockerfile/interoptest

tools/buildgen/generate_projects.sh is used to regenerate files from templates.

yongni replied:
I see. need to dig into more.  Thanks for the info.

jtattermusch commented on tools/run_tests/dockerize/build_interop_image.sh:
```diff
@@ -37,6 +37,7 @@ set -x
 #  INTEROP_IMAGE - name of tag of the final interop image
 #  BASE_NAME - base name used to locate the base Dockerfile and build script
 #  TTY_FLAG - optional -t flag to make docker allocate tty
+#  RUNTIME - runtime setting. If specified, Dockerfile-$RUNTIME will be used.
```
I am not convinced this is useful. Right now our interop docker images have just one dimension - the name.  having grpc_interop_go17 and grpc_interop_go18 is equally powerful as having grpc_interop_go with two "runtime" flavors, but the latter is more complicated and requires more changes.
Also, introducing a concept of "runtime" is problematic because that can lead to confusion what counts as "runtime" vs what should already be a different language (and my past experience suggests people will be inconsistent in thinking one way or another).

yongni replied:
Just to help me understand, are you suggesting we flatten all the runtimes and fold it into languages?  Basically, treating all the go/grpc-matrix variants as separate languages?

jtattermusch replied:
Yes, I am suggesting to flatten the configurations into what you call "languages" - but it's just different dockerfiles.   What is a "language" and "runtime" is basically something that go/grpc-matrix can determine according to your needs.  It just need to know that what is under "Java" and "OpenJDK 8"  corresponds to docker image "grpc_interop_java_openjdk8".  Not a big deal, and you don't need to propagate these concepts all the way to the level of run_interop_tests.py (or at least not to the level of build_interop_image shell scripts).

yongni commented:
I dropped java opensdk due to outstanding issues.  Just added go 1.7 and 1.8 template and build the Dockerfile from running tools/buildgen/generate_projects.sh.

This is the minimum I need to move forward. PTAL.  Thanks Jan for your feedback.
